README.md
# blade_logic Cookbook

TODO: Enter the cookbook description here.

e.g.
This cookbook makes your favorite breakfast sandwich.

## Requirements

TODO: List your cookbook requirements. Be sure to include any requirements this cookbook has on platforms, libraries, other cookbooks, packages, operating systems, etc.

e.g.
### Platforms

- SandwichOS

### Chef

- Chef 12.0 or later

### Cookbooks

- `toaster` - blade_logic needs toaster to brown your bagel.

## Attributes

TODO: List your cookbook attributes here.

e.g.
### blade_logic::default

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['blade_logic']['bacon']</tt></td>
    <td>Boolean</td>
    <td>whether to include bacon</td>
    <td><tt>true</tt></td>
  </tr>
</table>

## Usage

### blade_logic::default

TODO: Write usage instructions for each cookbook.

e.g.
Just include `blade_logic` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[blade_logic]"
  ]
}
```

## Contributing

TODO: (optional) If this is a public cookbook, detail the process for contributing. If this is a private cookbook, remove this section.

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

## License and Authors

Authors: TODO: List authors



CHANGELOG.md
# blade_logic CHANGELOG

This file is used to list changes made in each version of the blade_logic cookbook.

## 0.1.0
- [your_name] - Initial release of blade_logic

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/unit/recipes/redhat_spec.rb
#
# Cookbook:: spec_test
# Spec:: default
#
# Copyright:: 2017, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'blade_logic::redhat' do
  context 'Validate supported installations' do
    before do
     stub_data_bag_item('artifacts', 'blade_logic').and_return(
#       'blade_logic' =>  {
    		'am1' => {
             'package' => 'https://am1.artifactory/RSCD86-SP1-LIN64.rpm',
             'checksum' => '<shasum256>',
			 'version' => '8.6SP1'
            },
           'aws_us_east_1' => {
             'package' => 'https://locb.customeralpha.org/appbincustom.rpm',
             'checksum' => '<shasum256>',
			 'version' => '8.6SP1'
            }
#        }
      )
    end
  platforms = {
       'redhat' => {
         'versions' => %w(6.6 7.3)
      }
    }
    platforms.each do |platform, components|
      components['versions'].each do |version|
        context "On #{platform} #{version}" do
    
#  context 'When all attributes are default, on an rhel 6.6, 7.3' do
#    let(:chef_run) do
#      # for a complete list of available platforms and versions see:
#      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
#      runner = ChefSpec::ServerRunner.new(platform: platform, version: version)
#      runner.converge(described_recipe)
    end
    before do
              Fauxhai.mock(platform: platform, version: version)
              # Node attributes
              node.normal['aig']['datacenter'] = 'am1'
            end
            let(:runner) do
              ChefSpec::SoloRunner.new(platform: platform, version: version)
            end
            let(:node) { runner.node }
            let(:chef_run) { runner.converge(described_recipe) }

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
	end
	
	it 'creates a remote_file with attributes' do
      expect(chef_run).to create_remote_file("/var/chef/cache/blade_logic_client/RSCD86-SP1-LIN64.rpm").with(owner: 'root',group: 'root',mode: '0750' )
    end
	
	it 'installs a package when specifying the identity attribute' do
      expect(chef_run).to install_package('rscd')
    end
	
	it 'creates a template with attributes' do
      expect(chef_run).to create_template('/usr/lib/rsc/users.local').with(source: 'users.local.erb', owner: 'root',group: 'root',mode: '0644' )
	end
	
	it 'enables a service with an explicit action' do
     expect(chef_run).to enable_service('rscd')
    end
	
	it 'returns an error when datacenter attribute is not set' do
      node.normal['aig']['datacenter'] = nil 
     expect { chef_run }.to raise_error(ArgumentError, 'The chosen location is not supported or the blade_logic binary information not found.')
    end
	
	 it 'creates a directory with attributes' do
    %w(/var/chef /var/chef/cache /var/chef/cache/blade_logic_client).each do |dir_name|
	 expect(chef_run).to create_directory("#{dir_name}").with(
      user:   'root',
      group:  'root',
	  mode: '0755'
    )
	end
	end
	#it 'returns an error when chosen location is not found' do
     #node.normal['aig']['datacenter'] = 'bad_location'
     #expect { chef_run }.to raise_error(ArgumentError, 'The chosen location is not supported or the blade_logic binary information not found.')
    #end

    it 'returns an error when data_bag (\'artifacts\',\'blade_logic\') is not found' do
     stub_data_bag_item('artifacts', 'blade_logic').and_return(nil)
     expect { chef_run }.to raise_error(ArgumentError, 'Unable to locate the DataBagItem("artifacts","blade_logic")')
    end

    
	
end
end
end
  context 'Validate unsupported platforms' do
    platforms = {
      'windows' => {
        'versions' => %w(2012 2012r2)
      },
	  'ubuntu' => {
         'versions' => %w(14.04 16.04)
	  }
    }
    platforms.each do |platform, components|
      components['versions'].each do |version|
        context "On #{platform} #{version}" do
          context 'When all attributes are default' do
            before do
              Fauxhai.mock(platform: platform, version: version)
            end
            let(:chef_run) do
              ChefSpec::SoloRunner.new(platform: platform, version: version) do |node|
                # Node attributes
              end.converge(described_recipe)
            end

            it 'raises an exception' do
              expect { chef_run }.to raise_error(ArgumentError, "ERROR: Unsupported Operating system #{platform}. Please run this cookbook on RHEL systems only!!!")
            end
          end

end
end
end
end

end

test/smoke/default/default_test.rb
# # encoding: utf-8

# Inspec test for recipe spec_test::default

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

#unless os.windows? && os.ubuntu?
  # This is an example test, replace with your own test.
#  describe user('root'), :skip do
#    it { should exist }
#  end
#end

# This is an example test, replace it with your own test.
#describe port(80), :skip do
#  it { should_not be_listening }
#end
%w(/var/chef /var/chef/cache /var/chef/cache/blade_logic_client).each do |dir_name|
	describe file("#{dir_name}") do
		its('type') { should eq :directory }
		it { should be_directory }
	end
end

describe file('/var/chef/cache/blade_logic_client/RSCD86-SP1-LIN64.rpm') do
	its('type') { should eq :file }
	it { should be_file }
end

describe file('/usr/lib/rsc/users.local') do
	its('type') { should eq :file }
	it { should be_file }
end

	
describe package('rscd') do
  it { should be_installed }
#  its('version') { should eq '1.9.5' }
end


describe service('rscd') do
  it { should be_running }
  # it { should be_monitored.by("monit") }
end


spec/unit/recipes/redhat_spec - Copy.rb
#
# Cookbook:: spec_test
# Spec:: default
#
# Copyright:: 2017, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'blade_logic::redhat' do
  context 'Validate supported installations' do
    before do
      stub_data_bag_item('artifacts', 'blade_logic').and_return(
        {
    		'am1' => {
             'package' => 'https://am1.artifactory/RSCD86-SP1-LIN64.rpm',
             'checksum' => '<shasum256>',
			 'version' => '8.6SP1'
            },
           'aws_us_east_1' => {
             'package' => 'https://locb.customeralpha.org/appbincustom.rpm',
             'checksum' => '<shasum256>',
			 'version' => '8.6SP1'
            }
        }
      )
    end
    platforms = {
       'redhat' => {
         'versions' => %w(6.6 7.3)
      }
    }
    platforms.each do |platform, components|
      components['versions'].each do |version|
        context "On #{platform} #{version}" do
          context 'When all attributes are default' do
            before do
              Fauxhai.mock(platform: platform, version: version)
              # Node attributes
              node.normal['aig']['datacenter'] = 'am1'
			  blade_logic_artifact = stub_data_bag_item('artifacts', 'blade_logic')
			  dc_pref = node['aig']['datacenter']
              dc_pref_child_key = stub_data_bag_item('blade_logic_artifact', 'dc_pref')
              binary_name = stub_data_bag_item('dc_pref_child_key', 'package')
#              binary_name = stub_data_bag_item('dc_pref_child_key', 'package').split('/')[-1]

            end
            let(:runner) do
              ChefSpec::SoloRunner.new(platform: platform, version: version)
            end
            let(:node) { runner.node }
            let(:chef_run) { runner.converge(described_recipe) }

            it 'converges successfully' do
                 expect { chef_run }.to_not raise_error
              %w(/var /var/chef /var/chef/cache /var/chef/cache/blade_logic_client).each do |dir_name|
                 expect(chef_run).to create_directory(dir_name).with(owner: 'root', group: 'root', mode: '0755')
             end
			    expect(chef_run).to create_remote_file("/var/chef/cache/blade_logic_client/#{binary_name}").with(source: stub_data_bag_item('dc_pref_child_key', 'package'))
                expect(chef_run).to install_package("::File.basename(#{binary_name}, '.*')" ).with(source: "/var/chef/cache/blade_logic_client/#{binary_name}")
                expect(chef_run).to create_template('/usr/lib/rsc/users.local').with( user: 'root', group: 'root', mode: '0644' )

            end

            
        end
      end
    end
  end
  
end
end

recipes/redhat.rb
#
# Cookbook:: blade_logic
# Recipe:: redhat
#
# Copyright:: 2017, The Authors, All Rights Reserved.


raise ArgumentError,"ERROR: Unsupported Operating system #{node['platform'] }. Please run this cookbook on RHEL systems only!!!" if node['platform_family'] != 'rhel'

#blade_logic_artifact = data_bag_item('artifacts', 'blade_logic')
begin
  blade_logic_artifact = data_bag_item('artifacts', 'blade_logic')
  raise if blade_logic_artifact.nil?
rescue
  blade_logic_artifact = {} 
  raise ArgumentError, 'Unable to locate the DataBagItem("artifacts","blade_logic")'
end

dc_pref = node['aig']['datacenter']
raise ArgumentError, 'The chosen location is not supported or the blade_logic binary information not found.' if dc_pref.nil? 

#data_bag_item( 'artifacts', 'blade_logic' ).each do |dc_code|
#raise ArgumentError, 'The chosen location is not supported or the blade_logic binary information not found.' if dc_pref != dc_code 
#end


#dc_pref_child_key = blade_logic_artifact['dc_pref']
#binary_name = dc_pref_child_key['package']
#binary_name = blade_logic_artifact['dc_pref']['package']
binary_name = (blade_logic_artifact[node['aig']['datacenter']]['package']).split('/')[-1]

%w(/var/chef /var/chef/cache /var/chef/cache/blade_logic_client).each do |dir_name|
directory "#{dir_name}" do
		owner 'root'
		group 'root'
		mode '0755'
	action :create
	end
end
remote_file "/var/chef/cache/blade_logic_client/#{binary_name}" do
   source "https://am1.artifactory/#{binary_name}"
#  source "http://mirror.centos.org/centos/7/os/x86_64/Packages/#{binary_name}"
  #checksum 'checksum'
  owner 'root'
  group 'root'
  mode '0750'
  action :create
end

package 'Install rscd' do
  package_name  'rscd'             
  source "/var/chef/cache/blade_logic_client/#{binary_name}"
  action :install
end


template '/usr/lib/rsc/users.local' do
  source 'users.local.erb'
  owner 'root'
  group 'root'
  mode '0644'
end

service "rscd" do
    action :enable
end